text/2022-11-08-font-compiler-ir.md

```diff
@@ -0,0 +1,190 @@
+instance and subset it down as they see fit.
+
+The design should facilitate parallel execution (see
+[Simon’s explorations](https://github.com/googlefonts/oxidize/blob/main/text/2022-05-10-parallel-font-compile-experiments.md#what-discrete-operations-are-available))
+and incremental compilation.
+
+We should be able to optionally capture IR. IR should support human friendly (textual) and fast/compact (binary) representation (shamelessly copying ... I mean
+inspired by ... [LLVM IR](https://llvm.org/docs/Reference.html#llvm-ir), hopefully leaning on serde).
+
+## Potential future goals
+
+Enable translation from `sourceA → IR → sourceB` so we can modernise sources in an automated manner (credit @cmyr for this idea).
+
+## Non-goals
+
+Lossless `source → IR → source`.
+
+# Basic structure
+
+```
+Source → IR → font binary
+            ^ one ir2font implementation
+       ^ many source2ir implementations
+```
+
+Source is one of
+
+- .glyphs
+- UFO (including many w/design space)
+- .ttx
+- nanoemoji config
+
+In future, sources could additionally include:
+
+- binary font
+  - Some contend if we support ttx we effectively support binary fonts
+- Unknown Future Design Format
+
+For every source we will have a dedicated conversion to IR, e.g. `glyphs2ir`, `ufo2ir`, etc in https://github.com/googlefonts/fontmake-rs.
+
+Let’s take `ufo2ir` as a concrete example, it would:
+
+- Depend on a UFO parser, likely https://github.com/linebender/norad
+  - Aside: norad uses quick-xml, we should try to stick to that elsewhere
+- Depend on the IR library
+
+# IR
+
+## Font sources
+
+Rough outline of common source structures
+
+UFO:
+
+- Designspace file maps UFO(s) into design space
+- UFO font has layers
+- Layers have glyphs
+
+Glyphs:
+
+- Font has glyphs
+- Glyphs have one layer per master
+- Plus user defined layers
+- Plus "special" layers that have attributes
+  - For example, "brace" layers to define intermediate masters for specific glyphs
+
+## Sketch
+
+Based on [ufo3](https://unifiedfontobject.org/versions/ufo3/), [babelfont-rs](https://github.com/simoncozens/babelfont-rs) and discussion in this repo. The
+structures sketched here in pseudo-Rust (based on multiple reports a similar format was helpful for colr). We deliberately have no concept of "master", we
+instead have a set of variable glyphs whose definition locations need not be identical. The examples are not exhaustive of all fields, they merely aspire to
+convey the general idea of a variable first, sparse first, structure broken apart in alignment with units of work for a parallel backend:
+
+```rust
+// data common to the entire designspace
+// written to disk in one file
+// some of https://unifiedfontobject.org/versions/ufo3/fontinfo.plist/ 
+struct Font
+  upem
+  axes: Vec<Axis>
+
+// TODO: do we really need "discrete" axes? - designspace 5.0
```

Comment:
Yes, because a DSv5 can capture family-wide information like STAT that would have to be duplicated into continuous design spaces.

Reply:
That example is very helpful, thank you.

Reply:
Inlined in https://github.com/googlefonts/oxidize/commit/cb8153c3cf2028a6700a42951f533719294d938b